mmaction/core/evaluation/eval_detection.py

```diff
@@ -0,0 +1,249 @@
+                label_lst.append(activity_index[ann['label']])
+
+        ground_truth = pd.DataFrame({
+            'video-id': video_lst,
+            't-start': t_start_lst,
+            't-end': t_end_lst,
+            'label': label_lst
+        })
+        return ground_truth, activity_index
+
+    def _import_prediction(self, prediction_filename):
+        """Reads prediction file, returns the prediction instances.
+
+        Parameters
+        ----------
+        prediction_filename : str
+            Full path to the prediction json file.
+
+        Outputs
+        -------
+        prediction : df
+            Data frame containing the prediction instances.
+        """
+        with open(prediction_filename, 'r') as fobj:
+            data = json.load(fobj)
+        # Read predictions.
+        video_lst, t_start_lst, t_end_lst = [], [], []
+        label_lst, score_lst = [], []
+        for videoid, v in data['results'].items():
+            for result in v:
+                label = self.activity_index[result['label']]
+                video_lst.append(videoid)
+                t_start_lst.append(float(result['segment'][0]))
+                t_end_lst.append(float(result['segment'][1]))
+                label_lst.append(label)
+                score_lst.append(result['score'])
+        prediction = pd.DataFrame({
+            'video-id': video_lst,
+            't-start': t_start_lst,
+            't-end': t_end_lst,
+            'label': label_lst,
+            'score': score_lst
+        })
+        return prediction
+
+    def _get_predictions_with_label(self, prediction_by_label, label_name,
+                                    cidx):
+        """Get all predicitons of the given label.
+
+        Return empty DataFrame if there is no predcitions with the given label.
+        """
+        if cidx in prediction_by_label.groups.keys():
+            return prediction_by_label.get_group(cidx).reset_index(drop=True)
+        else:
+            print('Warning: No predictions of label \'%s\' were provdied.' %
+                  label_name)
+            return pd.DataFrame()
+
+    def wrapper_compute_average_precision(self):
+        """Computes average precision for each class."""
+        ap = np.zeros((len(self.tiou_thresholds), len(self.activity_index)))
+
+        # Adaptation to query faster
+        ground_truth_by_label = self.ground_truth.groupby('label')
+        prediction_by_label = self.prediction.groupby('label')
+
+        results = Parallel(n_jobs=len(self.activity_index))(
+            delayed(compute_average_precision_detection)(
+                ground_truth=ground_truth_by_label.get_group(cidx).reset_index(
+                    drop=True),
+                prediction=self._get_predictions_with_label(
+                    prediction_by_label, label_name, cidx),
+                tiou_thresholds=self.tiou_thresholds,
+            ) for label_name, cidx in list(self.activity_index.items()))
```

Comment:
for label_name, class_idx in self.activity_index.items()

Reply:
code updated